MultiServer Shooter Tutorial - Docs Fixes (#742)
spatialos/UnrealGDK: SpatialGDK/Documentation/content/get-started/tutorial.md

![Toolbar]({{assetRoot}}assets/screen-grabs/toolbar/schema-button.png)<br/>
_Image: On the GDK toolbar in the Unreal Editor select **Schema**_<br/>
1. Select [**Snapshot**]({{urlRoot}}/content/generating-a-snapshot) to generate a snapshot (a representation of the state of the SpatialOS world) which will be used to start the deployment.<br/>
![Toolbar]({{assetRoot}}assets/screen-grabs/toolbar/snapshot-button.png)<br/>
_Image: On the GDK toolbar in the Unreal Editor select **Snapshot**_<br/>

### Deploy the project locally

In this section you’ll run a [local deployment](https://docs.improbable.io/reference/latest/shared/glossary#local-deployment) of the project. As the name suggests, local deployments run on your development machine (you will [cloud deploy](https://docs.improbable.io/reference/latest/shared/glossary#cloud-deployment) later in this tutorial).

1. In a text editor, open `UnrealGDKThirdPersonShooter\spatial\default_launch.json` and in the `layer_configurations` section set `rows` to `2`. This instructs SpatialOS that you will be launching two server-workers.

> 1. In the Unreal Editor, in the Unreal toolbar, open the **Play** drop-down menu.<br/>

Reviewer: These should all be 1.

2. Under **Multiplayer Options**, enter the number of players as **2**
3. Enter the number of servers as **2**
4. Ensure the box next to **Run Dedicated Server** is checked<br/>
![]({{assetRoot}}assets/set-up-template/template-multiplayer-options.png)<br/>
_Image: The Unreal Engine **Play** drop-down menu, with **Multiplayer Options** and **New Editor Window (PIE)** highlighted_<br/>
1. In Unreal Editor, in the SpatialOS GDK toolbar, select **Start**, the green play icon. This will open a terminal window and run the [`spatial local launch`](https://docs.improbable.io/reference/latest/shared/spatial-cli/spatial-local-launch#spatial-local-launch) command, which starts the [SpatialOS Runtime](https://docs.improbable.io/reference/latest/shared/glossary#the-runtime).<br/>
![Toolbar]({{assetRoot}}assets/set-up-template/template-start.png)<br/>
_Image: On the GDK toolbar in the Unreal Editor select **Start**_<br/>
6. It's ready when you see `SpatialOS ready. Access the inspector at http://localhost:21000/inspector`.

> 7. From the Unreal Editor toolbar, click **Play** to run the game. This starts two headless server-workers and two [client-workers](https://docs.improbable.io/reference/latest/shared/glossary#client-worker) locally.

Reviewer 2: I've not seen the term 'headless server-worker' anywhere else in the docs. One for @ElleEss maybe? Think we might need to explain what we mean by headless.

Reviewer 3: I think "dedicated server" is the more common term for Unreal.

Author: Thanks both
Will defer to @ElleEss

Reviewer 4: I have never heard of this. 
What does it mean? Who wrote it? I have a very bright light I can shine in their face to interrogate.

Reviewer 4: Also, `select **Play**` is the MMoS standard (but that's less of an issue).

Author: Not sure who introduced it here but we've used the term before. Spatialos runs game engines in headless mode, described for instance here https://docs.improbable.io/reference/13.6/shared/byoe/introduction#requirements-for-engine-integration

> 7. From the Unreal Editor toolbar, click **Play** to run the game. This starts two server-worker instances and two [client-worker](https://docs.improbable.io/reference/latest/shared/glossary#client-worker) instances locally, in your Unreal Editor.

Reviewer 4: ```suggestion
7. From the Unreal Editor toolbar, click **Play** to run the game. This starts two SpatialOS server-worker instances and two SpatialOS [client-worker](https://docs.improbable.io/reference/latest/shared/glossary#client-worker) instances locally, in your Unreal Editor.
<br/>The two server-worker instances are acting as two Unreal servers and the two client-worker instances are acting as two Unreal game clients (two game players).
```

1. In the Unreal Editor, in the Unreal toolbar, open the **Play** drop-down menu.<br/>
2. Under **Multiplayer Options**, enter the number of players as **2**
3. Enter the number of servers as **2**
4. Ensure the box next to **Run Dedicated Server** is checked<br/>
![]({{assetRoot}}assets/set-up-template/template-multiplayer-options.png)<br/>
_Image: The Unreal Engine **Play** drop-down menu, with **Multiplayer Options** and **New Editor Window (PIE)** highlighted_<br/>
1. In the Unreal Editor, in the SpatialOS GDK toolbar, select **Start** (the green play icon). This opens a terminal window and runs the [`spatial local launch`](https://docs.improbable.io/reference/latest/shared/spatial-cli/spatial-local-launch#spatial-local-launch) command, which starts the [SpatialOS Runtime](https://docs.improbable.io/reference/latest/shared/glossary#the-runtime).<br/>
![Toolbar]({{assetRoot}}assets/set-up-template/template-start.png)<br/>
_Image: On the GDK toolbar in the Unreal Editor select **Start**_<br/>
6. It's ready when you see `SpatialOS ready. Access the inspector at http://localhost:21000/inspector`.

> 7. From the Unreal Editor toolbar, click **Play** to run the game. This starts two SpatialOS server-worker instances and two SpatialOS [client-worker](https://docs.improbable.io/reference/latest/shared/glossary#client-worker) instances locally, in your Unreal Editor.

Reviewer 4: ```suggestion
7. From the Unreal Editor toolbar, select **Play** to run the game. This starts two SpatialOS server-worker instances and two SpatialOS [client-worker](https://docs.improbable.io/reference/latest/shared/glossary#client-worker) instances locally, in your Unreal Editor.
```

> 7. From the Unreal Editor toolbar, select **Play** to run the game. This starts two SpatialOS server-worker instances and two SpatialOS [client-worker](https://docs.improbable.io/reference/latest/shared/glossary#client-worker) instances locally, in your Unreal Editor.

Reviewer 4: ```suggestion
7. From the Unreal Editor toolbar, select **Play** to run the game. This starts two SpatialOS server-worker instances and two SpatialOS client-worker instances locally, in your Unreal Editor.
<br/>(You can find out about workers in the [glossary](https://docs.improbable.io/unreal/alpha/content/glossary#workers).)
```

> <br/>The two server-worker instances are acting as two Unreal servers and the two client-worker instances are acting as two Unreal game clients (two game players).

Reviewer 4: ```suggestion
<br/>The two server-worker instances are acting as two Unreal servers and the two client-worker instances are acting as two Unreal game clients (as would be used by two game players).
<br/>(You can find out about workers in the [glossary](https://docs.improbable.io/unreal/alpha/content/glossary#workers).)
```

3. Enter the number of servers as **2**
4. Ensure the box next to **Run Dedicated Server** is checked<br/>
![]({{assetRoot}}assets/set-up-template/template-multiplayer-options.png)<br/>
_Image: The Unreal Engine **Play** drop-down menu, with **Multiplayer Options** and **New Editor Window (PIE)** highlighted_<br/>
1. In the Unreal Editor, in the SpatialOS GDK toolbar, select **Start** (the green play icon). This opens a terminal window and runs the [`spatial local launch`](https://docs.improbable.io/reference/latest/shared/spatial-cli/spatial-local-launch#spatial-local-launch) command, which starts the [SpatialOS Runtime](https://docs.improbable.io/reference/latest/shared/glossary#the-runtime).<br/>
![Toolbar]({{assetRoot}}assets/set-up-template/template-start.png)<br/>
_Image: On the GDK toolbar in the Unreal Editor select **Start**_<br/>
6. It's ready when you see `SpatialOS ready. Access the inspector at http://localhost:21000/inspector`.
7. From the Unreal Editor toolbar, select **Play** to run the game. This starts two SpatialOS server-worker instances and two SpatialOS client-worker instances locally, in your Unreal Editor.

> <br/>(You can find out about workers in the [glossary](https://docs.improbable.io/unreal/alpha/content/glossary#workers).)

Reviewer 4: ```suggestion
```

### View your SpatialOS world in the Inspector

![]({{assetRoot}}assets/tutorial/inspector-two-workers.png)

[The Inspector]({{urlRoot}}/content/glossary#inspector) provides a real-time view of what is happening in your [SpatialOS world]({{urlRoot}}/content/glossary#game-world). It’s a powerful tool for monitoring and debugging both during development and when your game is live in production. Let’s use the Inspector to visualise the areas that each of our server-workers have [authority]({{urlRoot}}/content/glossary#authority) (that is, read and write access) over.

1. Access the inspector at [http://localhost:21000/inspector](http://localhost:21000/inspector).

> 1. In the **View** tab, click the checkboxes next to both of the **UnrealWorkers**. In the top right section of the viewport, select Show me: **Authority / Interest**. This will cause the Inspector to display the areas that the server-workers have authority over as two coloured zones.

Reviewer 5: ```suggestion
1. In the **View** tab, click the checkboxes next to both of the **UnrealWorkers**.
1. In the top right section of the viewport, in the “Show me” box, select **Authority / interest**.<br>
This will cause the Inspector to display the areas that each server-worker has authority over as two coloured zones.
```

Reviewer 2: ```suggestion
1. In the **View** tab, select the checkboxes next to both of the **UnrealWorkers**. 
1. In the top right section of the viewport, select Show me: **Authority / Interest**. 
This makes the inspector display the areas that the server-workers have authority over as two coloured zones.
```

### View your SpatialOS world in the Inspector

> ![]({{assetRoot}}assets/tutorial/inspector-two-workers.png)

Reviewer 4: ```suggestion
![]({{assetRoot}}assets/tutorial/inspector-two-workers.png)

_Image: A local Inspector showing two server-worker instances (two Unreal servers) managing your game _<br/>
```

> ![]({{assetRoot}}assets/tutorial/inspector-two-workers.png)

Reviewer 4: OK, I guessed from the context what it looks like.

[The Inspector]({{urlRoot}}/content/glossary#inspector) provides a real-time view of what is happening in your [SpatialOS world]({{urlRoot}}/content/glossary#game-world). It’s a powerful tool for monitoring and debugging both during development and when your game is live in production. Let’s use the Inspector to visualise the areas that each of our server-workers have [authority]({{urlRoot}}/content/glossary#authority) (that is, read and write access) over.

1. Access the inspector at [http://localhost:21000/inspector](http://localhost:21000/inspector).
1. In the **View** tab, click the checkboxes next to both of the **UnrealWorkers**.

> 1. In the top right section of the viewport, in the “Show me” box, select **Authority / interest**.<br>

Reviewer 2: It does say "show me" 

![ShowMe](https://user-images.githubusercontent.com/24454135/54599620-e5754980-4a32-11e9-9eeb-d5098c756142.png)

Might be worth adding a screenshot.

> 1. In the top right section of the viewport, in the “Show me” box, select **Authority / interest**.<br>

Reviewer 4: Still not sure on a viewport. I am going to edit this as best it can be, merge and then look at adding this image.

### View your SpatialOS world in the Inspector

![]({{assetRoot}}assets/tutorial/inspector-two-workers.png)

> _Image: A local Inspector showing two server-worker instances (two Unreal servers) managing your game _<br/>

Reviewer 4: ```suggestion
_Image: A local Inspector showing two server-worker instances (two Unreal servers) managing your game_<br/>
```

1. In Unreal Editor, in the SpatialOS GDK toolbar, select **Start**. It's ready when you see `SpatialOS ready. Access the inspector at [http://localhost:21000/inspector]()`.
1. From the Unreal Editor toolbar, click **Play** to run the game.
2. Using the Inspector to track the location of your two players, notice that you can now shoot across servers and cause damage (provided the two players are on different teams!)

![]({{assetRoot}}assets/tutorial/shooting-across-boundaries.gif)<br/>

> *Image: Players Running and Shooting across two Unreal Servers*

Reviewer 2: <3

1. In Unreal Editor, in the SpatialOS GDK toolbar, select **Start**. It's ready when you see `SpatialOS ready. Access the inspector at [http://localhost:21000/inspector]()`.
1. From the Unreal Editor toolbar, click **Play** to run the game.

> 2. Using the Inspector to track the location of your two players, notice that you can now shoot across servers and cause damage (provided the two players are on different teams!)

Author: Fair - each worker is running an unreal server, so it is correct to say either IMO but we need to standardise

> 2. Using the Inspector to track the location of your two players, notice that you can now shoot across servers and cause damage (provided the two players are on different teams!)

Reviewer 4: ```suggestion
2. Using the Inspector to track the location of your two players, notice that you can now shoot between two Unreal servers and cause damage across their boundaries (provided the two players are on different teams!).
```

> 2. Using the Inspector to track the location of your two players, notice that you can now shoot across servers and cause damage (provided the two players are on different teams!)

Reviewer 4: I amended the text above to make it clear that, in this context a SpatialOS server-worker instance is synonymous with an Unreal server and a SpatialOS client-worker instance is synonymous with a game client (a game player).
This means we can use the familiar terms "Unreal server" and "game client" at the end of this text.

![]({{assetRoot}}assets/tutorial/shooting-across-boundaries.gif)<br/>

> *Image: Players Running and Shooting across two Unreal Servers*

Reviewer 4: Aren't they server-worker instances? Isn't it disingenious to use "server"? Needs thought.

> *Image: Players Running and Shooting across two Unreal Servers*

Reviewer 4: ```suggestion
*Image: Players running and shooting between two Unreal Servers*
```

1. In Unreal Editor, in the SpatialOS GDK toolbar, select **Start**. It's ready when you see `SpatialOS ready. Access the inspector at [http://localhost:21000/inspector]()`.
1. From the Unreal Editor toolbar, click **Play** to run the game.
2. Using the Inspector to track the location of your two players, notice that you can now shoot between two Unreal servers and cause damage across their boundaries (provided the two players are on different teams!).

![]({{assetRoot}}assets/tutorial/shooting-across-boundaries.gif)<br/>
*Image: Players running and shooting between two Unreal Servers*

> Now that you're free of the single-server paradigm, have a think about the huge, seamless multiplayer worlds you can build and host using the Unreal GDK.

Reviewer 4: ```suggestion
Now that you're free of the single-server paradigm, have a think about the huge, seamless multiplayer worlds you can build and host using the Unreal GDK.
<br/>
<br/>
-------------
2019-03-15 Page updated with limited editorial review
```

> 

Reviewer 4: <br/>
-------------
2019-03-15 Page updated with limited editorial review

An assembly is what’s created when you run `BuildWorker.bat`. They’re .zip files that contains all the files that your game uses when running in the cloud.

1. In a terminal window, change directory to the root directory of the Third Person Shooter repository.

> 2. Build a server-worker assembly by running: `Game\Plugins\UnrealGDK\SpatialGDK\Build\Scripts\BuildWorker.bat ThirdPersonShooterServer Linux Development ThirdPersonShooter.uproject`

Reviewer 5: ```suggestion
1. Build a server-worker assembly by running: `Game\Plugins\UnrealGDK\SpatialGDK\Build\Scripts\BuildWorker.bat ThirdPersonShooterServer Linux Development ThirdPersonShooter.uproject`
```

> 3. Build a client-worker assembly by running: `Game\Plugins\UnrealGDK\SpatialGDK\Build\Scripts\BuildWorker.bat ThirdPersonShooter Win64 Development ThirdPersonShooter.uproject`

Reviewer 5: ```suggestion
1. Build a client-worker assembly by running: `Game\Plugins\UnrealGDK\SpatialGDK\Build\Scripts\BuildWorker.bat ThirdPersonShooter Win64 Development ThirdPersonShooter.uproject`
```
